File: freud/py/telemetry/prompt_budget.py
#!/usr/bin/env python3
"""Prompt budget allocator telemetry aggregator.

Port of freud/scripts/prompt-budget-telemetry.sh — parses JSONL events of type
``prompt_budget_allocation`` and prints fallback/degradation counters with tuning hints.
"""

from __future__ import annotations

import argparse
import json
import sys
from collections import Counter
from pathlib import Path

from freud.py.common import pct, repo_root


def _load_events(path: Path, event_type: str) -> list[dict]:
    """Load JSONL file and return events matching *event_type*."""
    events = []
    for line in path.read_text(encoding="utf-8", errors="replace").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        if obj.get("type") == event_type:
            events.append(obj)
    return events


def prompt_budget_telemetry(input_path: str) -> None:
    """Aggregate and print prompt budget telemetry from *input_path*."""
    path = Path(input_path)
    if not path.is_file():
        print(f"Event log not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    events = _load_events(path, "prompt_budget_allocation")
    total = len(events)

    if total == 0:
        print(f"No prompt_budget_allocation events found in: {input_path}")
        return

    fallback_count = sum(1 for e in events if e.get("data", {}).get("single_message_fallback") is True)
    floor_violation_events = sum(1 for e in events if (e.get("data", {}).get("floor_violation_count") or 0) > 0)
    dropped_sections_total = sum(e.get("data", {}).get("dropped_section_count", 0) for e in events)

    costs = [e.get("data", {}).get("allocated_total_cost", 0) for e in events]
    floors = [e.get("data", {}).get("reserved_floor_cost", 0) for e in events]
    avg_cost = sum(costs) / len(costs) if costs else 0
    avg_floor = sum(floors) / len(floors) if floors else 0

    print("Prompt Budget Telemetry")
    print(f"source: {input_path}")
    print()
    print("Totals")
    print(f"- allocations: {total}")
    print(f"- single_message_fallback: {fallback_count} ({pct(fallback_count, total)}%)")
    print(f"- floor_violation_events: {floor_violation_events} ({pct(floor_violation_events, total)}%)")
    print(f"- dropped_sections_total: {dropped_sections_total}")
    print(f"- avg_allocated_total_cost: {avg_cost:.2f}")
    print(f"- avg_reserved_floor_cost: {avg_floor:.2f}")
    print()

    # Breakdown by call_site
    print("Breakdown by call_site")
    site_counts: Counter[str] = Counter()
    for e in events:
        site = e.get("data", {}).get("call_site") or "unknown"
        site_counts[site] += 1
    for site, count in site_counts.most_common():
        print(f"- {site}: {count}")
    print()

    # Breakdown by degradation_path
    print("Breakdown by degradation_path")
    deg_counts: Counter[str] = Counter()
    for e in events:
        deg = e.get("data", {}).get("degradation_path") or "none"
        deg_counts[deg] += 1
    for deg, count in deg_counts.most_common():
        print(f"- {deg}: {count}")
    print()

    # Fallback rate by call_site
    print("Fallback rate by call_site")
    site_events: dict[str, list[dict]] = {}
    for e in events:
        site = e.get("data", {}).get("call_site") or "unknown"
        site_events.setdefault(site, []).append(e)
    for site in sorted(site_events, key=lambda s: -len(site_events[s])):
        evts = site_events[site]
        t = len(evts)
        fb = sum(1 for e in evts if e.get("data", {}).get("single_message_fallback") is True)
        fv = sum(1 for e in evts if (e.get("data", {}).get("floor_violation_count") or 0) > 0)
        print(f"- {site}: total={t}, fallback={fb}, floor_violations={fv}")

    print()
    print("Tuning Hints")
    if fallback_count > 0:
        print("- single-message fallback occurred: reduce required floors or increase max prompt budget for affected call sites.")
    if floor_violation_events > 0:
        print("- floor violations occurred: required floor reservation exceeds budget in some prompts; inspect degradation_path and band usage.")
    if dropped_sections_total > 0:
        print("- sections were dropped: verify optional/context blocks are ordered and banded by true criticality.")
    if fallback_count == 0 and floor_violation_events == 0:
        print("- no severe prompt pressure observed in this sample.")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Aggregate prompt budget telemetry from JSONL events.",
        usage="freud/scripts/prompt-budget-telemetry.sh [events_jsonl]",
    )
    parser.add_argument("events_jsonl", nargs="?",
                        default=str(repo_root() / ".psyke/logs/latest-events.jsonl"),
                        help="Path to events JSONL file")
    args = parser.parse_args()

    prompt_budget_telemetry(args.events_jsonl)


if __name__ == "__main__":
    main()

File: freud/internal/analysis/telemetry/prompt_budget.go
package telemetry

import (
	"fmt"
	"sort"

	"github.com/atomitl/neopsyke/freud/internal/analysis"
)

// PromptBudgetTelemetry aggregates prompt budget allocation stats from a JSONL events file.
func PromptBudgetTelemetry(inputPath string) error {
	if !analysis.FileExists(inputPath) {
		return fmt.Errorf("event log not found: %s", inputPath)
	}

	events, err := analysis.LoadJSONLEvents(inputPath, "prompt_budget_allocation")
	if err != nil {
		return err
	}

	total := len(events)
	if total == 0 {
		fmt.Printf("No prompt_budget_allocation events found in: %s\n", inputPath)
		return nil
	}

	fallbackCount := 0
	floorViolationEvents := 0
	droppedSectionsTotal := 0
	var costs, floors []float64

	for _, e := range events {
		data := analysis.GetData(e)
		if fb, ok := analysis.GetBool(data, "single_message_fallback"); ok && fb {
			fallbackCount++
		}
		if fv, ok := analysis.GetInt(data, "floor_violation_count"); ok && fv > 0 {
			floorViolationEvents++
		}
		if ds, ok := analysis.GetInt(data, "dropped_section_count"); ok {
			droppedSectionsTotal += ds
		}
		if c, ok := data["allocated_total_cost"].(float64); ok {
			costs = append(costs, c)
		}
		if f, ok := data["reserved_floor_cost"].(float64); ok {
			floors = append(floors, f)
		}
	}

	avgCost := avg(costs)
	avgFloor := avg(floors)

	fmt.Println("Prompt Budget Telemetry")
	fmt.Printf("source: %s\n\n", inputPath)
	fmt.Println("Totals")
	fmt.Printf("- allocations: %d\n", total)
	fmt.Printf("- single_message_fallback: %d (%s%%)\n", fallbackCount, analysis.Pct(fallbackCount, total))
	fmt.Printf("- floor_violation_events: %d (%s%%)\n", floorViolationEvents, analysis.Pct(floorViolationEvents, total))
	fmt.Printf("- dropped_sections_total: %d\n", droppedSectionsTotal)
	fmt.Printf("- avg_allocated_total_cost: %.2f\n", avgCost)
	fmt.Printf("- avg_reserved_floor_cost: %.2f\n\n", avgFloor)

	// Breakdown by call_site
	fmt.Println("Breakdown by call_site")
	siteCounts := countBy(events, "call_site")
	for _, kv := range sortedCounts(siteCounts) {
		fmt.Printf("- %s: %d\n", kv.key, kv.count)
	}
	fmt.Println()

	// Breakdown by degradation_path
	fmt.Println("Breakdown by degradation_path")
	degCounts := countBy(events, "degradation_path")
	for _, kv := range sortedCounts(degCounts) {
		fmt.Printf("- %s: %d\n", kv.key, kv.count)
	}
	fmt.Println()

	// Fallback rate by call_site
	fmt.Println("Fallback rate by call_site")
	siteEvents := groupBy(events, "call_site")
	siteKeys := make([]string, 0, len(siteEvents))
	for k := range siteEvents {
		siteKeys = append(siteKeys, k)
	}
	sort.Slice(siteKeys, func(i, j int) bool {
		return len(siteEvents[siteKeys[i]]) > len(siteEvents[siteKeys[j]])
	})
	for _, site := range siteKeys {
		evts := siteEvents[site]
		t := len(evts)
		fb := 0
		fv := 0
		for _, e := range evts {
			data := analysis.GetData(e)
			if v, ok := analysis.GetBool(data, "single_message_fallback"); ok && v {
				fb++
			}
			if v, ok := analysis.GetInt(data, "floor_violation_count"); ok && v > 0 {
				fv++
			}
		}
		fmt.Printf("- %s: total=%d, fallback=%d, floor_violations=%d\n", site, t, fb, fv)
	}
	fmt.Println()

	fmt.Println("Tuning Hints")
	if fallbackCount > 0 {
		fmt.Println("- single-message fallback occurred: reduce required floors or increase max prompt budget for affected call sites.")
	}
	if floorViolationEvents > 0 {
		fmt.Println("- floor violations occurred: required floor reservation exceeds budget in some prompts; inspect degradation_path and band usage.")
	}
	if droppedSectionsTotal > 0 {
		fmt.Println("- sections were dropped: verify optional/context blocks are ordered and banded by true criticality.")
	}
	if fallbackCount == 0 && floorViolationEvents == 0 {
		fmt.Println("- no severe prompt pressure observed in this sample.")
	}

	return nil
}

func avg(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range vals {
		sum += v
	}
	return sum / float64(len(vals))
}

type kv struct {
	key   string
	count int
}

func countBy(events []map[string]interface{}, field string) map[string]int {
	counts := map[string]int{}
	for _, e := range events {
		data := analysis.GetData(e)
		val := analysis.GetString(data, field)
		if val == "" {
			val = "none"
		}
		counts[val]++
	}
	return counts
}

func groupBy(events []map[string]interface{}, field string) map[string][]map[string]interface{} {
	groups := map[string][]map[string]interface{}{}
	for _, e := range events {
		data := analysis.GetData(e)
		val := analysis.GetString(data, field)
		if val == "" {
			val = "unknown"
		}
		groups[val] = append(groups[val], e)
	}
	return groups
}

func sortedCounts(m map[string]int) []kv {
	var kvs []kv
	for k, v := range m {
		kvs = append(kvs, kv{k, v})
	}
	sort.Slice(kvs, func(i, j int) bool {
		return kvs[i].count > kvs[j].count
	})
	return kvs
}
